packages/editor-ui/src/plugins/codemirror/completions/utils.ts

```diff
@@ -1,14 +1,24 @@
 import { NODE_TYPES_EXCLUDED_FROM_AUTOCOMPLETION } from '@/components/CodeNodeEditor/constants';
+import { SPLIT_IN_BATCHES_NODE_TYPE } from '@/constants';
 import { useWorkflowsStore } from '@/stores/workflows';
+import { resolveParameter } from '@/mixins/workflowHelpers';
+import { useNDVStore } from '@/stores/ndv';
+import type { Completion, CompletionContext } from '@codemirror/autocomplete';
 
-export function autocompletableNodeNames() {
-	return useWorkflowsStore()
-		.allNodes.filter((node) => !NODE_TYPES_EXCLUDED_FROM_AUTOCOMPLETION.includes(node.type))
-		.map((node) => node.name);
+/**
+ * Split user input into base (to resolve) and tail (to filter).
+ */
+export function splitBaseTail(userInput: string): [string, string] {
+	const parts = userInput.split('.');
+	const tail = parts.pop() ?? '';
+
+	return [parts.join('.'), tail];
 }
 
-export const longestCommonPrefix = (strings: string[]) => {
-	if (strings.length === 0) return '';
+export function longestCommonPrefix(...strings: string[]) {
+	if (strings.length < 2) {
+		throw new Error('Expected at least two strings');
```

Comment:
i18n?

Reply:
This is an internal utility, not user visible. The check is only to prevent misuse.